sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,277 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Utilities for ``FileSystem`` implementations."""
+
+import abc
+import io
+import os
+
+__all__ = ['Downloader', 'Uploader', 'DownloaderStream', 'UploaderStream',
+           'PipeStream']
+
+
+class Downloader(object):
+  """Download interface for a single file.
+
+  Implementations should support random access reads.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def size(self):
+    """Size of file to download."""
+
+  @abc.abstractmethod
+  def get_range(self, start, end):
+    """Retrieve a given byte range from this download, inclusive.
+
+    Range must be in this form:
+      0 <= start <= end: Fetch the bytes from start to end.
+
+    Args:
+      start: (int) Initial byte offset.
+      end: (int) Final byte offset, inclusive.
+
+    Returns:
+      (string) A buffer containing the requested data.
+    """
+
+
+class Uploader(object):
+  """Upload interface for a single file."""
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def last_error(self):
+    """Last error encountered for this instance."""
```

Comment:
<!--thread_id:cc_164892461_t; commit:79815e6df1d49c271a6c44e37fb2a3c46be0aed5; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> In the docstring I mean, for future filesystem implementors.

<!--section:body-->
After further consideration, I think this property is safe to remove.